### PR TITLE
Adds ussd send on pin

### PR DESCRIPTION
**Pull request by mango-habanero**
**Describe the Pull Request**

This bug changes the send token flow of the ussd send token process. It removes the need to confirm what a user bought and initiates the sending of the actual token on entry of the user pin.

https://github.com/GrassrootsEconomics/CIC-Docs/projects/1#card-35821324

### CODE REVIEW

File: app/server/locale/ussd.sw.yml [hunk 1]
```diff
@@ -43,16 +43,14 @@ sw:
...
         10. Onyesha chaguzi zilizopita
     send_token_pin_authorization:
-      first: CON Weka nambari ya siri.
-        0. Nyuma
-      retry:
+      first: |
+        CON Tuma %{transaction_amount} %{token_name} kwa %{recipient_phone}.
+                Weka nambari ya siri kuthibitisha.
+                0. Nyuma
+                0. Nyuma
```
Comment (nolash) [inline app/server/locale/ussd.sw.yml]: 2 x 0?

File: app/server/utils/ussd/kenya_ussd_processor.py [hunk 2]
```diff
@@ -130,19 +130,25 @@ def custom_display_text(menu: UssdMenu, ussd_session: UssdSession) -> str:
...
             transaction_amount = ussd_session.get_data('transaction_amount')
-            transaction_reason = ussd_session.get_data('transaction_reason_i18n')
-            return i18n_for(
-                user, menu.display_key,
-                recipient_phone=recipient_phone,
-                token_name=token.symbol,
-                transaction_amount=cents_to_dollars(transaction_amount),
-                transaction_reason=transaction_reason
-            )
+            if user.failed_pin_attempts is not None and user.failed_pin_attempts > 0:
```
Comment (nolash) [inline app/server/utils/ussd/kenya_ussd_processor.py]: What should happen at `user.failed_pin_attempts == 0` ? is there code to catch that?

File: test/functional/app/api/test_ussd_api.py [hunk 3]
```diff
@@ -131,11 +114,11 @@ def mock_send_message(phone, message):
 
 
 def test_invalid_service_code(mocker, test_client,
-                                init_database, initialised_blockchain_network, init_seed):
+                              init_database, initialised_blockchain_network, init_seed):
 
     org = OrganisationFactory()
     sender = UserFactory(preferred_language="en", phone=make_kenyan_phone(phone()), first_name="Bob", last_name="Foo",
-        pin_hash=User.salt_hash_secret('0000'), default_organisation=org)
+                         pin_hash=User.salt_hash_secret('0000'), default_organisation=org)
```
Comment (nolash) [inline test/functional/app/api/test_ussd_api.py]: please args below each other or just one indent

File: test/helpers/ussd_tasker.py [hunk 4]
```diff
@@ -21,4 +26,4 @@ def inquire_balance(user: User):
 
     @staticmethod
     def fetch_user_exchange_rate(user: User):
-        pass
+        pass
```
Comment (nolash) [inline test/helpers/ussd_tasker.py:29]: do you have some auto-formatting on that messes with the last newline?

File: test/unit/app/utils/ussd/kenya/test_pin_state.py [hunk 5]
```diff
@@ -35,23 +37,23 @@
...
-     (new_pin_state, standard_user, "2222", "new_pin_confirmation"),
-     (new_pin_state, standard_user, "0000", "exit_invalid_pin"),
-     # new_pin_confirmation state tests
-     (new_pin_confirmation_state, standard_user, "2222", "complete"),
-     (new_pin_confirmation_state, standard_user, "1212", "exit_pin_mismatch"),
-     # opt_out_of_market_place_pin_authorization state tests
-     (opt_out_of_market_place_pin_authorization_state, pin_blocked_user, "1111",
-      "exit_pin_blocked")
- ])
+                         [
```
Comment (nolash) [inline test/unit/app/utils/ussd/kenya/test_pin_state.py]: hmm this looks like auto-formatting too. I'm not a fan.

File: test/unit/app/utils/ussd/kenya/test_pin_state.py [hunk 6]
```diff
@@ -71,12 +73,10 @@ def test_kenya_state_machine(test_client, init_database, user_factory, session_f
      (send_token_pin_authorization_state, pin_blocked_user, "1212", "exit_pin_blocked", 3, 3),
      (send_token_pin_authorization_state, standard_user, "1212",
       "send_token_pin_authorization", 1, 2),
-     (send_token_pin_authorization_state, standard_user, "0000",
-      "send_token_confirmation", 1, 0),
+     (send_token_pin_authorization_state, standard_user, "0000", "complete", 1, 0),
      (send_token_pin_authorization_state, standard_user, "1212",
```
Comment (nolash) [inline test/unit/app/utils/ussd/kenya/test_pin_state.py]: maybe all of them on one line then?

File: test/unit/app/utils/ussd/kenya/test_pin_state.py [hunk 7]
```diff
@@ -110,15 +110,16 @@ def test_kenya_state_machine(test_client, init_database, user_factory, session_f
  ])
 def test_authorize_pin(test_client, init_database, session_factory, user_factory, user_input, expected,
                        before_failed_pin_attempts, after_failed_pin_attempts):
-    session = session_factory()
-    user = user_factory(phone='+6140000000')
-    user.failed_pin_attempts = before_failed_pin_attempts
+    with db.session.no_autoflush:
```
Comment (nolash) [inline test/unit/app/utils/ussd/kenya/test_pin_state.py:99]: is there autoflush here?

**Review by nolash — CHANGES_REQUESTED** [review]

File: app/server/utils/ussd/kenya_ussd_processor.py (hunk 2 above)
Comment (nolash) [reply app/server/utils/ussd/kenya_ussd_processor.py]: actually that should be in the test also.

Comment (mango-habanero) [reply app/server/utils/ussd/kenya_ussd_processor.py]: > What should happen at `user.failed_pin_attempts == 0` ? is there code to catch that?

There's an else there yes ?
If the user.failed_pin_attempts == 0 then the else component in:

```
if user.failed_pin_attempts is not None and user.failed_pin_attempts > 0:
                return i18n_for(
                    user=user,
                    key="{}.{}".format(menu.display_key, 'retry'),
                    remaining_attempts=3 - user.failed_pin_attempts
                )
            else:
                return i18n_for(
                    user=user,
                    key="{}.{}".format(menu.display_key, 'first'),
                    transaction_amount=cents_to_dollars(transaction_amount),
                    token_name=token.symbol,
                    recipient_phone=recipient_phone
                )
```
will run and return the `send_token_pin_authorization` menu's `first` section

Comment (mango-habanero) [reply app/server/utils/ussd/kenya_ussd_processor.py]: > actually that should be in the test also.

The largest component of this ussd processor is the `custom_display_text` function which is already tested.
So even the scenario you've referred to is already implicitly tested.

File: test/functional/app/api/test_ussd_api.py (hunk 3 above)
Comment (mango-habanero) [reply test/functional/app/api/test_ussd_api.py]: The [`self-service-wallets-func-tests`](https://github.com/GrassrootsEconomics/SempoBlockchain/blob/philip/self-service-wallets-func-tests/test/functional/app/api/test_ussd_api.py) PR cleans all these lint issues.

File: test/helpers/ussd_tasker.py (hunk 4 above)
Comment (mango-habanero) [reply test/helpers/ussd_tasker.py:29]: I have pylint running for my scripts.

File: test/unit/app/utils/ussd/kenya/test_pin_state.py (hunk 7 above)
Comment (mango-habanero) [reply test/unit/app/utils/ussd/kenya/test_pin_state.py:99]: Yes there's a query invoked autoflush there

File: app/server/utils/ussd/kenya_ussd_processor.py (hunk 2 above)
Comment (nolash) [reply app/server/utils/ussd/kenya_ussd_processor.py]: Unless somehow the logical `and` between the two allow a different edge case to fall through...? Why for example isn't it `or` then?

File: test/unit/app/utils/ussd/kenya/test_pin_state.py (hunk 7 above)
Comment (nolash) [reply test/unit/app/utils/ussd/kenya/test_pin_state.py:99]:
```
$ grep autoflush test -r
test/unit/app/utils/ussd/kenya/test_kenya_ussd_processor.py:    with db.session.no_autoflush:
Binary file test/unit/app/utils/ussd/kenya/__pycache__/test_kenya_ussd_processor.cpython-36-PYTEST.pyc matches
```

I'm a bit sceptical here. _no other test_ has explicit autoflush check. Why is it needed in _this_ particular case please.

File: app/server/utils/ussd/kenya_ussd_processor.py (hunk 2 above)
Comment (mango-habanero) [reply app/server/utils/ussd/kenya_ussd_processor.py]: That's a really good catch. I think its overkill having it as it is. Maybe it should just be `user.failed_pin_attempts > 0:` the field not being none here is implicit

File: test/unit/app/utils/ussd/kenya/test_pin_state.py (hunk 7 above)
Comment (mango-habanero) [reply test/unit/app/utils/ussd/kenya/test_pin_state.py:99]: So when I run the test without the autoflush I run into an error along the lines of there being a user existing who bore the same number as the one that piece of code attempts to create.
The error then suggested that the cause of the error was a premature autoflush that was being invoked by some query within the flow of that piece of code. It the suggested that I run it without the autoflush to avoid the error and for sure it worked.

Comment (nolash) [reply test/unit/app/utils/ussd/kenya/test_pin_state.py:99]: the way the other tests work afaiu is that then you use a fixture to create a different user. (look for keyword "other" or "another" or something in the conftest files)

Whereas that's clearly not in any way elegant, I do not think we should introduce session-control methods that aren't introduced before. We are struggling with enough variance as it is in this system. Such can make it harder to clean things up once we make the pass to harden, sanitize and standardize the tests and fixtures.

Comment (mango-habanero) [reply test/unit/app/utils/ussd/kenya/test_pin_state.py:99]: So I revisited this today, I don't think there's a way around this when Sqlachemy runs that autoflush thing. 

> ```
> $ grep autoflush test -r
> test/unit/app/utils/ussd/kenya/test_kenya_ussd_processor.py:    with db.session.no_autoflush:
> Binary file test/unit/app/utils/ussd/kenya/__pycache__/test_kenya_ussd_processor.cpython-36-PYTEST.pyc matches
> ```
> 
> I'm a bit sceptical here. _no other test_ has explicit autoflush check. Why is it needed in _this_ particular case please.

Then I tried to find out if there were situations where it had been the current code base. I found one in:
`test/unit/app/utils/ussd/kenya/test_kenya_ussd_processor.py` in the `test_custom_display_text` test.
That aside, I am still looking for a way around but it looks like its one of those things where we can pass the no autoflush check in the session wide with:
`autoflush=false` or use what I've used there.

Comment (nolash) [reply test/unit/app/utils/ussd/kenya/test_pin_state.py:99]: > . I found one in:
test/unit/app/utils/ussd/kenya/test_kenya_ussd_processor.py in the test_custom_display_text 

which commit? I could not find that in your revision.

Comment (mango-habanero) [reply test/unit/app/utils/ussd/kenya/test_pin_state.py:99]: > > . I found one in:
> > test/unit/app/utils/ussd/kenya/test_kenya_ussd_processor.py in the test_custom_display_text
> 
> which commit? I could not find that in your revision.

https://github.com/GrassrootsEconomics/SempoBlockchain/commit/8517db39912709f00acf2b5d9d1a9741355c8a00

Comment (nolash) [reply test/unit/app/utils/ussd/kenya/test_pin_state.py:99]:
```
8517db399 (RubenGeo 2019-12-10 08:06:37 +0100 39) def test_custom_display_text(test_client, init_database, menu_name, language, expecteds,
8517db399 (RubenGeo 2019-12-10 08:06:37 +0100 40)                              real_at_idx, length, unexpected, menu_nr):
8517db399 (RubenGeo 2019-12-10 08:06:37 +0100 41)     with db.session.no_autoflush:
8517db399 (RubenGeo 2019-12-10 08:06:37 +0100 42)         start_state = UssdSessionFactory(state="menu_name")
```

Ok, thanks, I yield.

It seems like a hack, though, so we should make a point of weeding this out. I guess we are adding that to technical debt, unless you already found a way to do this cleanly and painlessly? We really need to get this functionality out.

**Review by nolash — COMMENTED** [review]
I cannot see other changes than just merge master in...?

**Comment by mango-habanero** [comment]
> I cannot see other changes than just merge master in...?

I am not sure I follow but here's a diff for master being merged in:
https://github.com/GrassrootsEconomics/SempoBlockchain/commit/4b4d697a2f7b84b2c5aea3a4d9d245181ca20d6a

**Review by nolash — APPROVED** [review]
